### PR TITLE
Switch BitTorrent analyzer to Zeek's standard regex engine

### LINKED ISSUE #1822: Switch BitTorrent analyzer to Zeek's standard regex engine
Turns out that `BitTorrentTracker.cc` uses POSIX regex functions (`#include <regex.h>`, `regcomp/regexec`) instead of Zeek's own engine. We should switch that over to remove that dependency and provide consistency.

### PR DESCRIPTION
**Pull request by avinal**
## Related Issue

- Fixes #1822

## What is Changed?

- Removes dependency on <regex.h>
- Replaces regex function with Zeek's standard regex functions
- Some replacements are workaround, may be improved later via an
appropiate API

Signed-off-by: Avinal Kumar <avinal.xlvii@gmail.com>

### CODE REVIEW

**Comment by avinal** [comment]
Hi @rsmmr, 
I have tried a workaround as you suggested. Please check and let me know if this will work. I am unable to determine it myself since the docs for Zeek's regex engine are not very descriptive 😅 .

File: src/analyzer/protocol/bittorrent/BitTorrentTracker.cc [hunk 1]
```diff
@@ -243,43 +243,67 @@ void BitTorrentTracker_Analyzer::DeliverWeird(const char* msg, bool orig)
 bool BitTorrentTracker_Analyzer::ParseRequest(char* line)
 	{
 	static bool initialized = false;
-	static regex_t r_get, r_get_end, r_hdr;
+	static RE_Matcher re_get("^GET[ \t]+");
+	static RE_Matcher re_get_end("[ \t]+HTTP/[0-9.]+$");
```
Comment (p-l-) [inline src/analyzer/protocol/bittorrent/BitTorrentTracker.cc:247]: This was already present in the existing code, but I think we should have this instead:
```suggestion
	static RE_Matcher re_get_end("[ \t]+HTTP/[0-9\.]+$");
```
Or am I missing something?

File: src/analyzer/protocol/bittorrent/BitTorrentTracker.cc [hunk 2]
```diff
@@ -344,12 +368,17 @@ void BitTorrentTracker_Analyzer::EmitRequest(void)
 bool BitTorrentTracker_Analyzer::ParseResponse(char* line)
 	{
 	static bool initialized = false;
-	static regex_t r_stat, r_hdr;
+	static RE_Matcher re_stat("^HTTP/[0123456789.]* ");
```
Comment (p-l-) [inline src/analyzer/protocol/bittorrent/BitTorrentTracker.cc:371]: Same here, plus you can use the `0-9` range as you properly suggested for `re_get_end` above
```suggestion
	static RE_Matcher re_stat("^HTTP/[0-9\.]* ");
```

**Review by rsmmr — COMMENTED** [review]
Nice job! The only issue I see is around the matching of `re_get_end`, see my notes. 

Also, adding a test would be good, looks like we don't have one for BitTorrent yet. Could you see if you find a trace with a simple BitTorrent connection somewhere and then add small btest ensuring the parsing still works?

File: src/analyzer/protocol/bittorrent/BitTorrentTracker.cc [hunk 3]
```diff
@@ -243,43 +243,67 @@ void BitTorrentTracker_Analyzer::DeliverWeird(const char* msg, bool orig)
...
+
+			// 1. Get the starting position of the match using MatchAnywhere()
+			// 2. Check if the match_end is greater than 0
+			// 3. If match_end is greater than 0, then there is a match
+			if ( (match_end = re_get_end.MatchAnywhere(line)) > 0 )
 				{
-				if ( match_end[0].rm_so <= match[0].rm_eo )
+				// 4. Get the length of the match using MatchPrefix()
+				// 5. Check if match_end is greater than len_get otherwise throw error
+				len_get_end = re_get_end.MatchPrefix(line);
```
[the remark this answers is not in the document]
Comment (avinal) [reply src/analyzer/protocol/bittorrent/BitTorrentTracker.cc:287]: Yeah, this was a lot confusing, I had to compare expected outputs from all the functions multiple, and this is what I came up with. I will fix it now.

**Comment by avinal** [comment]
> Also, adding a test would be good, looks like we don't have one for BitTorrent yet. Could you see if you find a trace with a simple BitTorrent connection somewhere and then add small btest ensuring the parsing still works?

Okay I will try to add a test too. A separate PR would be good?

File: src/analyzer/protocol/bittorrent/BitTorrentTracker.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/analyzer/protocol/bittorrent/BitTorrentTracker.cc:287]: I believe that should work if we do something like I suggested above, unless I'm missing something. If aren't sure how to proceed, I can give it a try if you want when I merge it (and after there's a test :-)

**Comment by rsmmr** [comment]
> Okay I will try to add a test too. A separate PR would be good?

You can add it to this PR, ideally as a separate commit coming first; then it's easy to check before/after the regex change.

File: src/analyzer/protocol/bittorrent/BitTorrentTracker.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (avinal) [reply src/analyzer/protocol/bittorrent/BitTorrentTracker.cc:287]: Ok let me write a test first, then it would be somewhat easy.

**Comment by rsmmr** [comment]
How is it looking?

**Comment by rsmmr** [comment]
I'm going to take a stab at wrapping this up.

**Comment by rsmmr** [comment]
Closing this in favor of #1910.

**Comment by avinal** [comment]
Hey @rsmmr, I apologize for not being able to complete this task. I was stuck with exams and submissions. Thanks a lot for wrapping this up.

**Pull request closed without merging**